uefi/src/main.rs

```diff
@@ -63,39 +69,77 @@ fn efi_main(image: Handle, st: SystemTable<Boot>) -> Status {
+    let mut kernel = load_kernel(image, &mut st, boot_mode);
+    if kernel.is_none() {
+        writeln!(
+            stdout,
+            "Failed to load kernel via {:?}, trying TFTP",
+            boot_mode
+        )
+        .unwrap();
+        // Try TFTP boot
+        boot_mode = BootMode::Tftp;
+        kernel = load_kernel(image, &mut st, boot_mode);
+    }
+    let kernel = kernel.expect("Failed to load kernel");
+    writeln!(stdout, "Trying to load ramdisk via {:?}", boot_mode).unwrap();
+    // Ramdisk must load from same source, or not at all.
+    let ramdisk = load_ramdisk(image, &mut st, boot_mode);
 
-    st.stdout().clear().unwrap();
     writeln!(
-        st.stdout(),
-        "UEFI bootloader started; trying to load kernel"
+        stdout,
+        "{}",
+        match ramdisk {
+            Some(_) => "Loaded ramdisk",
+            None => "Ramdisk not found.",
+        }
     )
     .unwrap();
 
-    let kernel = load_kernel(image, &st);
-
     let framebuffer = init_logger(&st, kernel.config);
-
-    // we no longer need the system table for printing panics
     unsafe {
         *SYSTEM_TABLE.get() = None;
     }
-
     log::info!("UEFI bootloader started");
     log::info!("Reading kernel and configuration from disk was successful");
     if let Some(framebuffer) = framebuffer {
         log::info!("Using framebuffer at {:#x}", framebuffer.addr);
     }
-
     let mmap_storage = {
-        let max_mmap_size =
-            st.boot_services().memory_map_size().map_size + 8 * mem::size_of::<MemoryDescriptor>();
-        let ptr = st
-            .boot_services()
-            .allocate_pool(MemoryType::LOADER_DATA, max_mmap_size)?;
-        unsafe { slice::from_raw_parts_mut(ptr, max_mmap_size) }
+        let mut memory_map_size = st.boot_services().memory_map_size();
+        let mut target_size = memory_map_size.map_size + (8 * memory_map_size.entry_size);
+        let mut storage: &mut [u8];
+        loop {
+            let ptr = st
+                .boot_services()
+                .allocate_pool(MemoryType::LOADER_DATA, target_size)
+                .expect("Failed to allocate memory for mmap storage");
+            storage = unsafe { slice::from_raw_parts_mut(ptr, target_size) };
+            if let Err(_) = st.boot_services().memory_map(storage) {
+                memory_map_size = st.boot_services().memory_map_size();
+                // By measuring the size here, we can find out exactly how much we need.
+                // We may hit this code twice, if the map allocation ends up spanning more pages.
+                let next_target_size = memory_map_size.map_size + (8 * memory_map_size.entry_size);
+                target_size = next_target_size;
+                st.boot_services()
+                    .free_pool(ptr)
+                    .expect("Failed to free temporary memory for memory map!");
+                continue;
+            }
+            break;
+        }
+        storage
```

Comment:
We can simplify this loop a bit:

```suggestion
        loop {
            let ptr = st
                .boot_services()
                .allocate_pool(MemoryType::LOADER_DATA, target_size)
                .expect("Failed to allocate memory for mmap storage");
            let storage = unsafe { slice::from_raw_parts_mut(ptr, target_size) };
            if st.boot_services().memory_map(storage).is_ok() {
                break storage;
            }
            // allocated memory region was not big enough -> free it again
            st.boot_services()
                .free_pool(ptr)
                .expect("Failed to free temporary memory for memory map!");

            // By measuring the size here, we can find out exactly how much we need.
            // We may hit this code twice, if the map allocation ends up spanning more pages.
            memory_map_size = st.boot_services().memory_map_size();

            let next_target_size = memory_map_size.map_size + (8 * memory_map_size.entry_size);
            target_size = next_target_size;
            
        }
```

Reply:
The measurement needs to be done prior to freeing the pool, this removes the need for the + 8 entries.